halo2_frontend/src/plonk/circuit.rs

```diff
@@ -10,9 +11,93 @@ pub mod expression;
 pub use constraint_system::*;
 pub use expression::*;
 
-// TODO: Bring ColumnType, Advice, Fixed, Instance and Any here, where Advice has a sealed phase
-// Keep a slim copy of those types in middleware.
-// https://github.com/privacy-scaling-explorations/halo2/issues/295
+/// A column type
+pub trait ColumnType:
+    'static + Sized + Copy + std::fmt::Debug + PartialEq + Eq + Into<Any>
+{
+    /// Return expression from cell
+    fn query_cell<F: Field>(&self, index: usize, at: Rotation) -> Expression<F>;
+}
+
+/// An advice column
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct Advice;
+
+/// A fixed column
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct Fixed;
+
+/// An instance column
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct Instance;
+
+impl ColumnType for Advice {
+    fn query_cell<F: Field>(&self, index: usize, at: Rotation) -> Expression<F> {
+        Expression::Advice(AdviceQuery {
+            index: None,
+            column_index: index,
+            rotation: at,
+            // phase: sealed::Phase(self.phase),
```

Comment:
Is this also leftover?

Reply:
Afaik, the `phase` infos are stored/moved here.
https://github.com/privacy-scaling-explorations/halo2/pull/300/files#diff-0e169fae906b996a10f9bd9a3ef3300f75683600792086d52fee37b409496527R134-R135

Reply:
> Whi is it commented out?? Do we have this info somewhere else??

Just like @duguorong009 said.  This info was already present but was duplicated in the `Advice` type.  From the PR description:
> - remove phase from `Advice`
>    - I noticed that the phase in the `Advice` was redundant because a column always exists in the context of a `ConstraintSystem` which has the `advice_column_phase` with that same information.  This removal simplifies the code.

The same applies to `AdviceQuery`, as it always lives in the context of a `ConstraintSystem`

Reply:
Removed leftover in https://github.com/privacy-scaling-explorations/halo2/pull/300/commits/e3acbd779df6c393070c1a43dd5640dd821d9ffe